src/devices/oregon_scientific.c

```diff
@@ -806,16 +826,37 @@ static int oregon_scientific_v3_decode(r_device *decoder, bitbuffer_t *bitbuffer
         return 1;
     }
     else if ((msg[0] == 0x20) || (msg[0] == 0x21) || (msg[0] == 0x22) || (msg[0] == 0x23) || (msg[0] == 0x24)) { // Owl CM160 Readings
+
         msg[0] = msg[0] & 0x0f;
         if (validate_os_checksum(decoder, msg, 22) != 0)
             return DECODE_FAIL_MIC;
-        float rawAmp = (msg[4] >> 4 << 8 | (msg[3] & 0x0f) << 4 | msg[3] >> 4);
-        unsigned short int ipower = (rawAmp /(0.27*230)*1000);
+
+        //channel add from https://github.com/magellannh/rtl-wx/blob/dbaf3924815903c47b230c65841d18b263421854/src/rtl-433fm-decode.c
+        int channel = msg[0]>>4;
+        int id = msg[1] & 0x0F;
+
+        //cm160 ipower rework & total energy added from https://github.com/magellannh/rtl-wx/blob/dbaf3924815903c47b230c65841d18b263421854/src/rtl-433fm-decode.c
+        unsigned int raw_ipower = cm160_power(msg); 
+        unsigned int ipower = (raw_ipower / (0.27*230) * 1000); //Assuming device is running in 230V country
+        //Alternate formula = (raw_ipower * 0.07) * 230 - https://github.com/cornetp/eagle-owl/blob/master/src/cm160.c
+
+        double raw_total_energy = cm160_total(msg);
+        double total_kWh = (raw_total_energy * 230 ) / 3600.0 / 1000.0;  //Assuming device is running in 230V country
```

Comment:
If `kWh = raw * 230 / 3600 / 1000` then `J = Ws = raw * 230`, so the unit is `Ws` or `J`, or actually `As = raw` which is strange but plausible.

Reply:
Fixed, this device measures amps, current and total.

Reply:
Makes sense. But then "total_amps" should be "total_ampsec" -- it is energy (A*s) not power (A).
And in data_make below it should be "current_A" (but this is not really A? There is a factor of 0.27?) and "total_As"

Reply:
Added unit A to current and total amps, there should be no need to indicate seconds as "One Ampere is defined as the current that flows with electric charge of one Coulomb **per second**"

Re: factor .27 ;  _this is above my pay grade_, that's how I found it ;-)  ; maybe the formulae can be improved but am not sure how the original author determined it.

Reply:
Ok, it's easy to confuse *per second* with *x seconds* ;) You will see *kmh* and know it's *per second*, i.e. *km/h*, but here I really meant *As* not *A/s* :)
An "instant flow of electric current" is `A` or like you found `C/s`. But an "accumulated amount of electric charge" is `C` wich is `A·s` :)
They are proportional with *power* and *energy (work)*, i.e. multiply with voltage and you get `W` and `Ws (=J) respectively.
In short it's not both "amps", they are very different, the later is "accumulated by time" -- work, the thing you actually pay for.

Reply:
Ok cool, so what must I change, starting to get confusing.

Reply:
Maybe indicate on the lines below what it should be:

                "current_A",        "Current Amps",         DATA_FORMAT,   "%d A", DATA_INT, current_amps,
                "total_A",          "Total Amps",           DATA_FORMAT,   "%d A", DATA_INT, (int)total_amps,

Reply:
These two values are returned by the machine un-manipulated to Amperage this current second and the cumulative Amperage (added every second) since the machine was factory reset.  Personally, I find the reading readings - Current Amps and Total Amps acceptable.  But if insisting I can change, just indicate.

Reply:
Sorry and thanks for sticking with this.
The `"total_A"` line should be `"total_As"` and `"%d As"`
But thinking about this you need to comment out both lines. As a rule we only output every value once, not converted to different formats. So
```
            // "current_A",        "Current Amps",         DATA_FORMAT,   "%d A", DATA_INT, current_amps,
            // "total_As",          "Total Amps",           DATA_FORMAT,   "%d As", DATA_INT, (int)total_amps,
```
As reference for future readers and changes. Sorry, again.

Reply:
All good, the part about hard coding the Voltage makes the solution less usable for countries where these devices are set to eg 110V. Just comment......

Reply:
Current output after remove the Amp readings

![image](https://user-images.githubusercontent.com/15066052/224574773-0ae9844d-29c6-4a86-bbc1-2d906a3bc1ca.png)

Reply:
Yes, that's a very strange thing about those meters. To count energy (or accurately show power) you will need to measure voltage. Pretty sure the device is doing that, the usual chips measure with "True RMS" -- they measure voltage very detailed, i.e. not only the effective ~230 V but the whole sine wave.

The `(0.27*230)*1000` is not correctly written, why would you *divide* by voltage?
The (same factor) of `0.07 * 230` is better, but it's just `*16.1` -- a random factor that likely was just estimated (linear regression?)
So maybe it is really just plain `*16` -- which is exactly 4 bits, so maybe we are missing those bits?
Then no 230 needed and it would always work, everywhere.

Strange about the energy though, we do need the 230 and the `*230/3600/1000` just looks perfect.
But then `3600/230` is also nearly exactly 16! A good guess would be that the field actually is not `As` but `Wh` already…

Seems all very much more likely than random conversions.

Reply:
Can you test if `current_watts = (current_amps * 16); // are we missing a nibble?`
and `total_kWh = total_amps / 16.0 / 1000.0; // should (msg[5]&0xf) belong to current_watts actually?`
work with accuracy? They seem much more probable.

Reply:
I tested, unfortunately, off by a couple of watts which is not consistent with the device's LCD value.  Nor am I convinced that the external device sending the 433 signals and clamping onto the power wire is aware of the volts.  During setup the volts are captured on the internal LCD device and it is not transmitted to the external clamp as the calculation of watt occur on the internal device.

Reply:
Oh, it's a contactless current-clamp only. I wasn't aware. Thanks for testing.

Reply:
> Oh, it's a contactless current-clamp only. I wasn't aware. Thanks for testing.

I have a second one on another property I will also verify the kWh calc with, I was able to reset it recently so the number is fairly low and visible on the LCD. Then if there's a discrepancy we can chat again.